doc/designs/thread-api.md

```diff
@@ -0,0 +1,305 @@
+
+This method can be dropped from MVP if it is deemed unnecessary.
+
+Manual Thread Method
+--------------------
+
+With this method, the application creates its own threads and donates them to
+OpenSSL by calling a function on those threads which does not return (unless
+later requested to do so).
+
+Example usage is demonstrated below:
+
+```c
+OSSL_thread_pool_cookie *cookie;
+
+{
+    /* error checking omitted for brevity */
+
+    cookie = OSSL_thread_pool_cookie_new();
+    OSSL_thread_pool_enter(cookie);
+    OSSL_thread_pool_cookie_free(cookie);
+    /* above function returns due to exit call below */
+}
+```
+
+If an application wishes to reclaim the thread from the thread pool, it
+may do so using the cookie:
+
+```c
+{
+    /* Causes OSSL_thread_pool_enter to return */
+    OSSL_thread_pool_exit(cookie);
+}
+```
+
+The cookie is created separately to avoid race conditions; the application must
+not call `OSSL_thread_pool_exit` before `OSSL_thread_pool_cookie_new` has
+returned.
+
+Because `OSSL_thread_pool_cookie_free` is synchronised, as an alternative usage
+model, another thread may simultaneously cause a thread to leave a thread pool
+and free its cookie:
+
+```c
+OSSL_thread_pool_cookie *cookie;
+
+/* thread pool thread */
+{
+    /* error checking omitted for brevity */
+
+    cookie = OSSL_thread_pool_cookie_new();
+    OSSL_thread_pool_enter(cookie);
+    /* above function returns due to cookie being freed */
+}
+
+/* another controlling thread */
+{
+    OSSL_thread_pool_cookie_free(cookie);
+}
+```
+
+The API is as follows:
+
+```c
+typedef struct ossl_thread_pool_cookie_st OSSL_THREAD_POOL_COOKIE;
+
+/*
+ * Creates a thread pool cookie.
+ */
+OSSL_THREAD_POOL_COOKIE *OSSL_THREAD_POOL_COOKIE_new(void);
+
+/*
+ * Frees a thread pool cookie.
+ *
+ * Calling this implicitly calls OSSL_thread_pool_exit_wait, therefore it is
+ * safe to call even when other threads have calls to OSSL_thread_pool_enter in
+ * progress, and will not return until those threads have exited
+ * OSSL_thread_pool_enter.
+ */
+void OSSL_THREAD_POOL_COOKIE_free(OSSL_THREAD_POOL_COOKIE *cookie);
+
+/*
+ * Resets a thread pool cookie which was passed to OSSL_thread_pool_exit so that
+ * it can be used again.
+ */
+void OSSL_THREAD_POOL_COOKIE_reset(OSSL_THREAD_POOL_COOKIE *cookie);
+
+/*
+ * Donates the executing thread to the OpenSSL thread pool. cookie must have
+ * been created by calling OSSL_THREAD_POOL_COOKIE_new.
+ *
+ * If the application desires to recover the thread at a later time,
+ * it can call OSSL_thread_pool_exit using the same cookie, which
+ * will cause this function to return.
+ *
+ * Returns 0 if entering the thread pool failed and 1 if the thread pool
+ * was entered successfully but the function subsequently returned
+ * due to a call to OSSL_thread_pool_exit.
+ */
+int OSSL_thread_pool_enter(OSSL_THREAD_POOL_COOKIE *cookie);
+
+/*
+ * Causes any in progress calls to OSSL_thread_pool_enter, or any future such
```

Comment:
Add a note about such calls might return after this one returns?
Otherwise, I don't see the difference between this and the waiting version.